Extend and ehance caching, add profiling

tndhagedorn opened the pull request:
Caching enhancements:
* Cache read values on `get`
* Cache the absence of a value when it is deleted/replaced
* On write, check cache to see if the type/value pair already exists in the database before writing
* Similar optimizations for `replace`, `delete`

Add opt-in profiling.  This can be used to see where an app is spending its time when using Vessel.

Add ability to pre-load all Vessel data into the cache.  This can have significant performance improvements depending on data access patterns (see profiling, above).

In general, many random reads are slower than one large read up-front. Re-writing the same data is also slower than reading everything up-front and using caching to elide unnecessary writes.

Update Robolectric dependency to 4.7 so that it works on Arm Macs (https://github.com/robolectric/robolectric/issues/6311)

Note that the caching is still not atomic with respect to the database operations.  This matches the existing implementation, and to limit scope this will be addressed in a follow-up PR.

tndhagedorn committed:
Extend caching

Caching enhancements:
* Cache read values
* Cache the absence of a value on deleted/replaced
* On write, check cache to see if the type/value pair already exists
  in the database before writing
* Similar optimizations for `replace`, `delete`

Add opt-in profiling.  This can be used to see where an app is spending
its time when using Vessel.

Add ability to preload all Vessel data into the cache.  This can
have significant performance improvements depending on data
access patterns (see profiling, above).

In general, many random reads are slower than one large read up-front.
Re-writing the same data is also slower than reading everything up-front
and using caching to elide equivalent writes.

tndhagedorn committed:
Simplify, clean up Profiler table code and unit test

malachid commented on build.gradle.kts:
```diff
@@ -23,6 +25,16 @@ allprojects {
         jcenter()
     }
+
+    tasks.withType(KotlinCompile::class.java).all {
```
I wonder if this syntax would be better
https://android.googlesource.com/platform/external/jetbrains/kotlin/+/refs/heads/mirror-goog-studio-master-dev/build.gradle.kts#310

malachid commented on vessel-runtime/src/main/java/com/textnow/android/vessel/Profiler.kt:
```diff
@@ -0,0 +1,358 @@
+ * A significant event that occurred when using a [Vessel] instance.
+ * Used to profile important events such as cache hits.
+ */
+enum class Event {
```
If this was a CacheHitEvent, with READ/WRITE/etc, you wouldn't need niceName()?

tndhagedorn replied:
It just keeps the summary output a bit nicer and shorter especially when dumping a lot of profiling info.

Instead of:

```
Spans
           span | count |  time
   READ_FROM_DB |     2 |    56
    WRITE_TO_DB |    16 |   474
        PRELOAD |     1 |    17

Events
          event | count
 CACHE_HIT_READ |   313
CACHE_HIT_WRITE |   128
```

You have:

```
Database Operations
           span | count |  time
   read |     2 |    56
    write |    16 |   474
        preload |     1 |    17

Cache Hits
          event | count
 read |   313
write |   128
```

tndhagedorn replied:
A full report looks something like:

```
Database I/O times, sorted by time spent
    span │ count │ time 
═════════╪═══════╪══════
   write │    15 │  947 
    read │    17 │  815 
 preload │     1 │   18 
  delete │     0 │    0 
 replace │     0 │    0 
   clear │     0 │    0 
─────────┼───────┼──────
         │    33 │ 1780 

Cache hits, sorted by hit count
   event │ count 
═════════╪═══════
    read │   202 
   write │   134 
  delete │     0 
 replace │     0 
─────────┼───────
         │   336 

Database I/O times, by thread/coroutine, sorted by time spent
      type │                        name │ write │ read │ preload │ delete │ replace │ clear │ total 
═══════════╪═════════════════════════════╪═══════╪══════╪═════════╪════════╪═════════╪═══════╪═══════
 COROUTINE │              Dispatchers.IO │   827 │  197 │       0 │      0 │       0 │     0 │  1024 
    THREAD │  DefaultDispatcher-worker-2 │     0 │  263 │      18 │      0 │       0 │     0 │   281 
    THREAD │  DefaultDispatcher-worker-7 │     0 │  181 │       0 │      0 │       0 │     0 │   181 
    THREAD │                   Thread-38 │     0 │  109 │       0 │      0 │       0 │     0 │   109 
    THREAD │  DefaultDispatcher-worker-1 │   105 │    0 │       0 │      0 │       0 │     0 │   105 
    THREAD │  DefaultDispatcher-worker-8 │     0 │   38 │       0 │      0 │       0 │     0 │    38 
 COROUTINE │         Dispatchers.Default │    15 │   15 │       0 │      0 │       0 │     0 │    30 
    THREAD │                        main │     0 │    7 │       0 │      0 │       0 │     0 │     7 
    THREAD │  DefaultDispatcher-worker-4 │     0 │    3 │       0 │      0 │       0 │     0 │     3 
    THREAD │  DefaultDispatcher-worker-3 │     0 │    2 │       0 │      0 │       0 │     0 │     2 
    THREAD │                   Thread-35 │     0 │    0 │       0 │      0 │       0 │     0 │     0 
    THREAD │                   Thread-55 │     0 │    0 │       0 │      0 │       0 │     0 │     0 
    THREAD │                   Thread-39 │     0 │    0 │       0 │      0 │       0 │     0 │     0 
    THREAD │ DefaultDispatcher-worker-12 │     0 │    0 │       0 │      0 │       0 │     0 │     0 
    THREAD │                   Thread-48 │     0 │    0 │       0 │      0 │       0 │     0 │     0 
    THREAD │  DefaultDispatcher-worker-5 │     0 │    0 │       0 │      0 │       0 │     0 │     0 
    THREAD │  DefaultDispatcher-worker-6 │     0 │    0 │       0 │      0 │       0 │     0 │     0 
───────────┼─────────────────────────────┼───────┼──────┼─────────┼────────┼─────────┼───────┼───────
           │                             │   947 │  815 │      18 │      0 │       0 │     0 │  1780 
```

malachid commented on vessel-runtime/src/main/java/com/textnow/android/vessel/Profiler.kt:
```diff
@@ -0,0 +1,358 @@
+ * Tracks data for one [Span] - how many times it was entered, and how much time (total) has been
+ * spent inside this span
+ */
+data class SpanData(val span: Span, var hitCount: Int = 0, var totalDurationMS: Long = 0L)
```
should have lower case 's' - ie: `totalDurationMs`

tndhagedorn replied:
I would have though `MS` given the [naming conventions](https://kotlinlang.org/docs/coding-conventions.html#choose-good-names)

> When using an acronym as part of a declaration name, capitalize it if it consists of two letters (IOStream); capitalize only the first letter if it is longer (XmlFormatter, HttpInputStream).

Golint for ex would force MS here.

malachid replied:
Hmm, that is interesting.
I was basing it on the syntax in the Kotlin SDK
https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-debug/kotlinx.coroutines.debug.junit5/-coroutines-timeout/test-timeout-ms.html?query=val%20testTimeoutMs:%20Long

tndhagedorn replied:
Yeah, that is interesting.  I'd suggest `MS` as it fits the "official" guidelines, but I'm fine with either since both seem to be a convention.

malachid commented on vessel-runtime/src/main/java/com/textnow/android/vessel/Profiler.kt:
```diff
@@ -0,0 +1,358 @@
+ * Helper to print text tables
+ */
+private fun table(
+    headings: List<Any>,
```
do these need to be `Any` since you know the format?

tndhagedorn replied:
`headings` can be `String`.  The others can be anything that would be shown in table form - Ints, Enums, etc (usage here may just be Int, but I don't see a need to constrain the function further even if it is just used here).  

I updated `headings` but left the rest.

malachid commented on vessel-runtime/src/main/java/com/textnow/android/vessel/Profiler.kt:
```diff
@@ -0,0 +1,358 @@
+            WorkerData(worker)
+        }
+
+        // Note:  synchronized is safe - and required - here:
```
multiline comments should use the `/** ... */` syntax
https://kotlinlang.org/docs/coding-conventions.html#documentation-comments

malachid commented on vessel-runtime/src/main/java/com/textnow/android/vessel/Vessel.kt:
```diff
@@ -44,6 +44,38 @@ interface Vessel {
+     * Many random reads can be slower than one larger read.  Similarly one large read is likely
+     * faster than re-writing many values.
+     */
+    suspend fun preload()
```
since the interface is changing, we'll want to mark this as a breaking change in case anyone had their own implementation.  don't forget to update the CHANGELOG.

malachid commented on vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt:
```diff
@@ -101,14 +122,65 @@ class VesselImpl(
+            dao.getAll()
+        }
+
+        preloadImpl(entities)
```
the time it takes to process each entity should be included in the time it takes to preload the db

tndhagedorn replied:
This should be inconsequential, but I did put it in just in case.

malachid commented on vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt:
```diff
@@ -101,14 +122,65 @@ class VesselImpl(
+            dao.getAllBlocking()
+        }
+
+        preloadImpl(entities)
```
same here

malachid commented on vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt:
```diff
@@ -101,14 +122,65 @@ class VesselImpl(
      * Convert a stored json string into the specified data type.
      */
-    private fun <T : Any> fromJson(value: String, type: KClass<T>): T? = gson.fromJson<T>(value, type.java)
+    private fun <T : Any> fromJson(value: String, type: KClass<T>): T? {
```
this change should be undone. idiomatic kotlin to use = instead of block with single line return

malachid commented on vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt:
```diff
@@ -285,21 +459,34 @@ class VesselImpl(
                     )
-                )
+                }
+                // Note - caching the result of the replace is safe, as any transactional Dao calls will throw on failure
```
multiline comment format

malachid commented on vessel-runtime/src/test/java/com/textnow/android/vessel/ProfilerTest.kt:
```diff
@@ -0,0 +1,186 @@
+import kotlin.random.Random
+
+/**
+ * Validate ability to mock Vessel when an in-memory db is not desired.
```
Test description is out of date

malachid commented on vessel-runtime/src/test/java/com/textnow/android/vessel/VesselCacheUsageTest.kt:
```diff
@@ -0,0 +1,295 @@
+    }
+}
+
+data class Data1(val field: Int = 1)
```
There are existing nested models that can be used for improved testing under src/test/java/com/textnow/android/vessel/model

malachid commented on vessel-runtime/src/test/java/com/textnow/android/vessel/VesselCacheUsageTest.kt:
```diff
@@ -0,0 +1,295 @@
+            return if (async) {
+                vessel.get(type)
+            } else {
+                vessel.getBlocking(type)
```
it does not look like the blocking versions are used below

malachid commented on vessel-runtime/src/test/java/com/textnow/android/vessel/VesselCacheUsageTest.kt:
```diff
@@ -0,0 +1,295 @@
+class ObserversVesselCacheUsageTest: BaseVesselTest<TestCache>(TestCache()) {
+    // Note:  This will deadlock certain transactional DAO methods, such as replace (but not set/delete)
+    @get:Rule()
+    val executor = InstantTaskExecutorRule()
```
You should be setting up the dispatchers. See FlowTest.

tndhagedorn replied:
Is this needed?  I guess I'm just not clear on what that adds to the test.

Ex, the following just uses runBlocking with no special dispatcher:

https://github.com/textnow/vessel/blob/a71de0dbefd0ee4f7a05b43d69c5fda0e55a7bbf/vessel-runtime/src/test/java/com/textnow/android/vessel/LiveDataTest.kt#L55

malachid replied:
Since some of your tests are relying on the main thread, it would be best to set it up as per
https://developer.android.com/kotlin/coroutines/test#setting-main-dispatcher

It's unclear if it would also address your deadlock.

tndhagedorn replied:
I wonder if this could be considered an instrumented test as it runs under Robolectric.
Same as your LiveDataTest.  If this test code can be kept simpler that would be ideal - I don't think the custom dispatcher is required.

The docs you linked indicate a dispatcher is only set up for local tests:
>Note: This only applies to local unit tests. You should not replace the Main dispatcher in [instrumented tests](https://developer.android.com/training/testing/instrumented-tests) where the real UI thread is available.

tndhagedorn committed:
Review comments

tndhagedorn committed:
Review comments

shawn-bailey commented on vessel-runtime/src/main/java/com/textnow/android/vessel/NoOpVessel.kt:
```diff
@@ -37,6 +37,11 @@ import kotlin.reflect.KClass
+
+    override fun preloadBlocking() { /* no-op */ }
+
+    override val profileData: ProfileData? = null
```
the spacing here isnt consistent with the rest of the class

shawn-bailey reviewed (commented):
Overall im ok with these changes. However I do have concerns around unchecked memory usage by loading the entire database at once. I dont see our internal use of vessel shrinking anytime soon and we would need some way to make sure reading everything at once doesnt spin out of control into another memory useage/db usage problem.

tndhagedorn commented:
> Overall im ok with these changes. However I do have concerns around unchecked memory usage by loading the entire database at once. I dont see our internal use of vessel shrinking anytime soon and we would need some way to make sure reading everything at once doesnt spin out of control into another memory useage/db usage problem.

Good point.

Current numbers are something like 16ms for a full load, on a low-end UMX phone, which currently saves about 10x the I/O time at startup by avoiding unnecessary reads/writes to disk.

So while I think we have a fair amount of headroom (for our current use case), I'm open to discussing how we can monitor this.  I have some ideas and I'm curious on your thoughts - let's sync offline.

tndhagedorn committed:
Review comment - fix spacing

a-ware commented on vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt:
```diff
@@ -158,12 +281,7 @@ class VesselImpl(
-                fromJson(entity, type.kotlin)
-            }
-        }
+        return getBlocking(type.kotlin)
```
Now that this is a one liner, it can use the `=` syntax instead of braces

a-ware commented on vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt:
```diff
@@ -201,11 +339,7 @@ class VesselImpl(
-            cache?.remove(typeName)
-            dao.deleteBlocking(typeName)
-        }
+        deleteBlocking(type.kotlin)
```
Same here for using the `=` syntax instead of braces

tndhagedorn replied:
Thanks, done!

a-ware commented on vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt:
```diff
@@ -328,12 +517,14 @@ class VesselImpl(
-                        }
+                .distinctUntilChanged()
+                .map {
+                    it?.data?.let { entity ->
+                        val data = fromJson(entity, type)
+                        cache?.set(typeName, data as Any)
+                        data
                     }
```
Optional, but you could use `also` here in place of the local val for these blocks. Use whatever you prefer for readability though
```suggestion
                    it?.data?.let { entity ->
                        fromJson(entity, type).also { data ->
                            cache?.set(typeName, data as Any)
                        }
                    }
```

tndhagedorn replied:
Thanks!  Good callout, and done!

a-ware reviewed (commented):
I haven't had a chance to test the changes yet, but they look good to me. I just found a few syntax nits, but nothing blocking

tndhagedorn committed:
Review comments, add timeout to preload API

Some code tweaks - review comments.

Add timeout to `preload/preloadBlocking`.  This can be used
to prevent preload time from growing too large as the database grows over time.

tndhagedorn commented:
> > Overall im ok with these changes. However I do have concerns around unchecked memory usage by loading the entire database at once. I dont see our internal use of vessel shrinking anytime soon and we would need some way to make sure reading everything at once doesnt spin out of control into another memory useage/db usage problem.
> 
> Good point.
> 
> Current numbers are something like 16ms for a full load, on a low-end UMX phone, which currently saves about 10x the I/O time at startup by avoiding unnecessary reads/writes to disk.
> 
> So while I think we have a fair amount of headroom (for our current use case), I'm open to discussing how we can monitor this. I have some ideas and I'm curious on your thoughts - let's sync offline.

Added a timeout to the `preload` APIs to address this issue.  See code/tests.

tndhagedorn committed:
Remove unused variable TAG from VesselImpl

shawn-bailey commented on vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt:
```diff
@@ -127,38 +127,81 @@ class VesselImpl(
+                val obj = if (data != null) fromJson(data, kclass) else nullValue
+                cache?.set(type, obj as Any)
+
+                timeoutMS?.let {
```
I think we should try to leverage coroutine timeouts here instead - while the docs show cancellation is still asynchronous it will still happen earlier than waiting for the entire span like you have here.  https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/with-timeout.html

tndhagedorn replied:
I don't think that will work here:
>The code that is executing inside the [block](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/with-timeout.html) is cancelled on timeout and the active or next invocation of the cancellable suspending function inside the block throws a [TimeoutCancellationException](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-timeout-cancellation-exception/index.html).

There are no suspend points in this function, so it won't cancel.

tndhagedorn replied:
hmnm... maybe not... digging.
I'm a little worried about the timeout happening anywhere at anytime if that is indeed possible.  
I'd prefer to use this approach and have explicit control when the timeout is happening.

tndhagedorn replied:
NM, I was right the first time.
You have to add suspension points to non-suspending code to make them cancellable, ex: https://kotlinlang.org/docs/cancellation-and-timeouts.html#making-computation-code-cancellable

Since both approaches are explicit, I'd prefer to keep this existing approach which can be common for the the blocking and non-blocking preload implementations.

a-ware commented on vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt:
```diff
@@ -51,44 +51,64 @@ import kotlin.reflect.KClass
+    private val allowMainThread: Boolean = false,
+    private val callback: VesselCallback? = null,
+    private val cache: VesselCache? = null,
+    private val profile: Boolean = false,
```
Missed adding a `@param` to the class header comment for the new `profile` param :]

tndhagedorn replied:
Good catch, will add.

a-ware reviewed (approved):
Looks good to me and replacing our app's Vessel dependency with this branch works fine.

I threw in a `preloadBlocking()` call right after we instantiate Vessel and didn't notice any issues on startup. I also dumped the profile summary into logs at a few places — pretty neat!

tndhagedorn commented:
> Looks good to me and replacing our app's Vessel dependency with this branch works fine.
> 
> I threw in a `preloadBlocking()` call right after we instantiate Vessel and didn't notice any issues on startup. I also dumped the profile summary into logs at a few places — pretty neat!

That's exactly what I did!  Saw a pretty big speedup in some cases

tndhagedorn committed:
Review comments - add @param for profile in VesselImpl

malachid commented on CHANGELOG.md:
```diff
@@ -7,6 +7,23 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [0.1.4] - 2023-06-09
```
So your changes just go under the "Unreleased" section.  The new version heading is created automatically during the release process, and items moved automatically.

malachid commented on vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt:
```diff
@@ -101,14 +122,104 @@ class VesselImpl(
      * Convert a stored json string into the specified data type.
      */
-    private fun <T : Any> fromJson(value: String, type: KClass<T>): T? = gson.fromJson<T>(value, type.java)
+    private fun <T : Any> fromJson(value: String, type: KClass<T>): T? = gson.fromJson(value, type.java)
```
IIRC the IDE said this generic parameter wasn't required, but we were getting exceptions without it

tndhagedorn replied:
At runtime?  How?  Generics are type erased.

malachid commented on vessel-runtime/src/test/java/com/textnow/android/vessel/BaseVesselTest.kt:
```diff
@@ -42,8 +42,8 @@ import org.koin.test.inject
  */
-abstract class BaseVesselTest(cache: VesselCache? = null) : KoinTest {
-    val vessel: Vessel by inject { parametersOf(cache) }
+abstract class  BaseVesselTest<T: VesselCache>(protected val cache: T? = null, protected val profile: Boolean = false, protected val inMemory: Boolean = true) : KoinTest {
```
this is getting long. we should split the params onto new lines

tndhagedorn committed:
Review comments

alinedubardfmf84 reviewed (commented):
Let's see
